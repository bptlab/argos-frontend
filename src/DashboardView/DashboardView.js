import React, {Component} from 'react';
import Header from './Header/Header.js';
import Diagram from './Diagram/Diagram.js';
import SearchBar from './SearchBar/SearchBar.js';
import ProductCardGrid from './ProductCardGrid/ProductCardGrid.js';
import Loader from '../Loader/Loader.js';

class DashboardView extends Component {
    constructor(props) {
        super(props);
        this.state = {
            searchText: '',
            products:   null,
            productFamilies: null,
            error:      null,
            excludedStates: []
        };
        // function binding
        this.handleSearchInput = this.handleSearchInput.bind(this);
        this.handleProductData = this.handleProductData.bind(this);
        this.handleProductFamilyData = this.handleProductFamilyData.bind(this);
        this.handleError = this.handleError.bind(this);
        this.handleExcludeStateInput = this.handleExcludeStateInput.bind(this);
    }
    
    handleExcludeStateInput(state) {
        const stateIndex = this.state.excludedStates.indexOf(state);
        if(stateIndex === -1) {
            this.setState({
                excludedStates: this.state.excludedStates.concat([state])
            });
        } else {
            const newStates = this.state.excludedStates;
            newStates.splice(stateIndex,1);
            this.setState({
                excludedStates: newStates
            });
        }
    }

    handleSearchInput(searchText) {
        this.setState({searchText: searchText});
    }
    
    componentDidMount() {
        this.fetchData();
        this.props.dataSource.notificationService.subscribe("Product", this.fetchData.bind(this));
    }

    componentWillUnmount() {
        this.props.dataSource.notificationService.unsubscribe("Product", this.fetchData.bind(this));
    }
    
    fetchData() {
        this.props.dataSource.fetchProducts(this.handleProductData, this.handleError);
        this.props.dataSource.fetchProductFamilies(this.handleProductFamilyData, this.handleError);
    }

    handleProductData(products) {
        this.setState({
            products: products
        });
    }

    handleProductFamilyData(productFamilies) {
        this.setState({
            productFamilies: productFamilies
        });
    }
    
    handleError(errorCode) {
        this.setState({
            error: errorCode
        });
    }
    
    render() {
        let component = (<Loader/> );
        if(this.state.error) {
            component = (
                <div className="critical-error">
                    <i className="fa fa-exclamation-triangle warning-sign" />
                    <p>{this.state.error}</p>
                </div>
            );
        } else if(this.state.products) {
            component = (
                <div>
                    <Header 
                        dataSource={this.props.dataSource}
                        dataSender={this.props.dataSender} />
                    <Diagram 
                        products={this.state.products}
                        onStateExcludeInput={this.handleExcludeStateInput} />
                    <SearchBar
                        onChangeSearchInput={this.handleSearchInput}
                        searchText={this.state.searchText}/>
                    <ProductCardGrid
                        productFamilies={this.state.productFamilies}
                        searchText={this.state.searchText}
                        excludedStates={this.state.excludedStates} />
                </div>
            );
        }
        return(
            <div className="dashboard-view">
                {component}
            </div>
        );
    } 
} 
export default DashboardView;