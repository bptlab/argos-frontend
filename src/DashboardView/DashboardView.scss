@import "Diagram/Diagram";
@import "Header/Header";
@import "ProductCardGrid/ProductCardGrid";
@import "SearchBar/SearchBar";
@import "Settings/Settings";