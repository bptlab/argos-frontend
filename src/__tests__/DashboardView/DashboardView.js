import React from 'react';
import DashboardView from '../../DashboardView/DashboardView.js';
import renderer from 'react-test-renderer';
import TestData from '../testData/products.js'
import ProductFamilyData from '../testData/frontend_productFamilies'
let instance, component, notificationService;
const fetchProducts = jest.fn();
const fetchProductFamilies = jest.fn();

beforeEach(() => {
    notificationService = {
        subscribe: jest.fn(),
        unsubscribe: jest.fn()
    };
    component = renderer.create(
        <DashboardView
            ref={(child) => {instance = child}}
            dataSource={{
                notificationService: notificationService,
                fetchProducts: fetchProducts,
                fetchProductFamilies: fetchProductFamilies
            }}/>
    );
    instance.handleProductFamilyData(ProductFamilyData.PRODUCTFAMILIES);
});

test("Correct drawing of DashboardView", () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
    expect(notificationService.subscribe).toBeCalled();
    expect(fetchProductFamilies).toBeCalledWith(instance.handleProductFamilyData, instance.handleError)
});

test("Error handling", () => {
    const errorCode = "An test-error occurred";
    instance.handleError(errorCode);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("Add a state to exclude from view", () => {
    instance.handleExcludeStateInput("RUNNING");
    expect(instance.state.excludedStates[0]).toEqual("RUNNING");
    instance.handleExcludeStateInput("RUNNING");
    expect(instance.state.excludedStates.length).toEqual(0);
});

test("Search in SearchBar", () => {
    instance.handleSearchInput("ProductDescription with Error");
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("Unmounting component", () => {
   instance.componentWillUnmount();
   expect(notificationService.subscribe).toBeCalled();
});