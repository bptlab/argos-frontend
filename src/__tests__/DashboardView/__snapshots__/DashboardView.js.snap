exports[`test Correct drawing of DashboardView 1`] = `
<div
  className="dashboard-view">
  <div>
    <header>
      <div
        className="container">
        <div
          className="row d-flex align-items-center">
          <div
            className="hidden-xs-down col-sm-1 d-flex justify-content-around">
            <a
              href="/#">
              <img
                alt="Overview page"
                className="product-icon"
                src="boiler.svg" />
            </a>
          </div>
          <div
            className="col-xs-11 col-sm-10 d-flex justify-content-around">
            <h1
              className="dashboard-title">
              Early Warning Dashboard
            </h1>
          </div>
          <div
            className="hidden-xs-down col-sm-1 d-flex justify-content-around">
            <div
              className="settings">
              <a
                className="nav-link"
                data-target="#settings-modal"
                data-toggle="modal"
                onClick={[Function]}>
                <i
                  className="settings-icon fa fa-bars" />
              </a>
              <div
                className="settings modal fade"
                id="settings-modal"
                role="dialog"
                tabIndex="-1">
                <div
                  className="modal-dialog modal-lg"
                  role="document">
                  <div
                    className="modal-content">
                    <div
                      className="modal-header">
                      <h5
                        className="modal-title"
                        id="settings-label">
                        Settings
                      </h5>
                      <button
                        className="close"
                        data-dismiss="modal"
                        type="button">
                        <span>
                          Ã—
                        </span>
                      </button>
                    </div>
                    <div
                      className="modal-body">
                      <div
                        aria-multiselectable="true"
                        id="accordion"
                        role="tablist">
                        <div
                          className="card">
                          <div
                            className="card-header"
                            id="headingOne"
                            role="tab">
                            <h5
                              className="mb-0">
                              <a
                                aria-controls="collapseOne"
                                aria-expanded="true"
                                data-parent="#accordion"
                                data-toggle="collapse"
                                href="#collapseOne">
                                Event-Types
                              </a>
                            </h5>
                          </div>
                          <div
                            aria-labelledby="headingOne"
                            className="collapse show"
                            id="collapseOne"
                            role="tabpanel">
                            <div
                              className="card-block">
                              <div
                                className="event-types-settings">
                                
                                <table
                                  className="table table-striped event-types-table">
                                  <thead>
                                    <tr>
                                      <th>
                                        #
                                      </th>
                                      <th>
                                        Name
                                      </th>
                                      <th>
                                        Query
                                      </th>
                                      <th
                                        className="center">
                                        Delete?
                                      </th>
                                    </tr>
                                  </thead>
                                  <tbody />
                                </table>
                                <p>
                                  To add a new Event-Type please refer to a Product and press the \"Plus\"-Button.
                                </p>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      className="modal-footer">
                      <button
                        className="btn btn-primary"
                        disabled="disabled"
                        onClick={[Function]}
                        type="button">
                        <div>
                          Exit
                        </div>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </header>
    <div
      className="container">
      <div
        className="diagram">
        <div
          className="row">
          <div
            className="col-xs-12 offset-sm-3 col-sm-6 offset-lg-4 col-lg-4 offset-uhd-5 col-uhd-2">
            <canvas
              id="pieChart" />
          </div>
        </div>
      </div>
    </div>
    <div
      className="container">
      <div
        className="searchArea">
        <div
          className="input-group">
          <span
            className="input-group-addon"
            id="search-addon">
            <i
              className="fa fa-search" />
          </span>
          <input
            className="form-control"
            onChange={[Function]}
            placeholder="Search..."
            type="text"
            value="" />
        </div>
      </div>
    </div>
    <div
      className="container">
      <div
        className="row">
        <div
          className="col-xs-12 col-sm-6 col-md-6 col-lg-3 col-uhd-2">
          <a
            href="/#/product/0"
            title="Details for Testproduct 5000 14E-17B">
            <div
              className="card ProductCard ProductCard_RUNNING">
              <h4
                className="card-title ProductCardTitle ProductCardTitle_RUNNING">
                Testproduct 5000 14E-17B
              </h4>
              <ul>
                <li
                  className="ProductCardInformationLine PCIL_serialNumber">
                  9830510416
                </li>
                <li
                  className="ProductCardInformationLine PCIL_quantity">
                  52189
                </li>
                <li
                  className="ProductCardInformationLine PCIL_date">
                  2016-05-07
                </li>
                <li
                  className="ProductCardInformationLine PCIL_errorQuantity">
                  84
                </li>
              </ul>
            </div>
          </a>
        </div>
        <div
          className="col-xs-12 col-sm-6 col-md-6 col-lg-3 col-uhd-2">
          <a
            href="/#/product/1"
            title="Details for Testproduct 5100 12E-15B">
            <div
              className="card ProductCard ProductCard_WARNING">
              <h4
                className="card-title ProductCardTitle ProductCardTitle_WARNING">
                Testproduct 5100 12E-15B
              </h4>
              <ul>
                <li
                  className="ProductCardInformationLine PCIL_serialNumber">
                  7840516417
                </li>
                <li
                  className="ProductCardInformationLine PCIL_quantity">
                  3215
                </li>
                <li
                  className="ProductCardInformationLine PCIL_date">
                  2016-03-01
                </li>
                <li
                  className="ProductCardInformationLine PCIL_errorQuantity">
                  54
                </li>
              </ul>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`test Error handling 1`] = `
<div
  className="dashboard-view">
  <div
    className="critical-error">
    <i
      className="fa fa-exclamation-triangle warning-sign" />
    <p>
      An test-error occurred
    </p>
  </div>
</div>
`;

exports[`test Search in SearchBar 1`] = `
<div
  className="dashboard-view">
  <div
    className="critical-error">
    <i
      className="fa fa-exclamation-triangle warning-sign" />
    <p>
      An test-error occurred
    </p>
  </div>
</div>
`;
